pkg/resource/logging-credentials/reconciler.go

```diff
@@ -0,0 +1,76 @@
+package loggingcredentials
+
+import (
+	"context"
+	"fmt"
+
+	loggedcluster "github.com/giantswarm/logging-operator/pkg/logged-cluster"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apimachineryerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Reconciler implements a reconciler.Interface to handle
+// Logging Credentials: store and maintain logging credentials
+type Reconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// ReconcileCreate ensures a secret exists for the given cluster.
+func (r *Reconciler) ReconcileCreate(ctx context.Context, lc loggedcluster.Interface) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info(fmt.Sprintf("loggingcredentials checking secret %s/%s", LoggingCredentialsSecretMeta(lc).Namespace, LoggingCredentialsSecretMeta(lc).Name))
+
+	var loggingCredentialsSecret v1.Secret
+
+	// Check if secrets exist / retrieve existing secret
+	err := r.Client.Get(ctx, types.NamespacedName{Name: LoggingCredentialsSecretMeta(lc).Name, Namespace: LoggingCredentialsSecretMeta(lc).Namespace}, &loggingCredentialsSecret)
+	if err != nil {
+		if apimachineryerrors.IsNotFound(err) {
+			logger.Info("loggingcredentials secret not found, initializing one")
+			// Create basic secret
+			loggingCredentialsSecret, err = GenerateLoggingCredentialsBasicSecret(lc)
+		}
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
+		}
+	}
+
+	// update the secret's contents if needed
+	secretUpdated := UpdateLoggingCredentials(&loggingCredentialsSecret)
+
+	// commit our changes
+	if secretUpdated {
+		logger.Info("loggingCredentials - Updating secret")
+		err = r.Client.Update(ctx, &loggingCredentialsSecret)
+		if apimachineryerrors.IsNotFound(err) {
+			logger.Info("loggingCredentials - Secret does not exist, creating it")
+			err = r.Client.Create(ctx, &loggingCredentialsSecret)
```

Comment:
As earlier, this is a odd way of doing error handling. Instead, I would explicitly check for the error
```
err = r.Client.Create(ctx, &loggingCredentialsSecret)
if err != nil {
	return ctrl.Result{}, errors.WithStack(err)
}
```

Reply:
Well, this one is a bit different:
- we try to update the secret
- if it fails because it does not exist (`IsNotFound`), we create it
- other errors are "real" errors and should fail
- If "create" fails, well it's a real failure too.

I updated this part of code in 3abf188c947eb66d903101e9b256dc450f3c12e0, maybe it makes more sense now?

Reply:
If you add more code after the `if apimachineryerrors.IsNotFound(err) {}` block, you may inadvertently overwrite the error and forget to check the error returned in `err = r.Client.Create(ctx, loggingCredentialsSecret)`. That's why I propose to explicitly add the error handling for that
```
err = r.Client.Create(ctx, &loggingCredentialsSecret)
if err != nil {
	return ctrl.Result{}, errors.WithStack(err)
}
```